### PR TITLE
Added the ability to export images from a prefix

### PR DESCRIPTION
**Pull request by gbenhaim**
1. Added the ability to export images from the current prefix,
   either as  a layered image or as a base image (Lago will create a
   new image which is a merge of the layer and its base).
   When exporting the image as a layer, the path to his parent will
   be replaced with the current directory, relative to the path of the
   exported image (in simple words, it expect that the parent will be with
   him in the same directory). All this logic can be found in export.py.

2. Added the ability to resolve the parent of a layered image
   (for now only one level, and only for "template_type=qcow2").
   Lago will check if the parent is in the cache, if not it will
   download it from the template repo that was provided, and then
   create a symlink from the layer's directory to the parent.

3. When using a file based disk, allow the user to choose between
   using the actual file, or let Lago make a copy of the disk.
   The default is to use the actual file, and it can be changed through
   the init file with specifying 'make_a_copy=true' under the disk
   section.

4. Added some utility functions to support the flow that was explained
   above.

5. Moved hash calculation function from templates.py to utils.py.

6. Added a verification to 'qemu-img convert' in template.py.

7. Added functional tests for the new feature.

Signed-off-by: gbenhaim <galbh2@gmail.com>

closes https://github.com/lago-project/lago/issues/160

### CODE REVIEW

File: lago/cmd.py [hunk 1]
```diff
@@ -271,6 +271,39 @@ def do_stop(prefix, vm_names, **kwargs):
     prefix.stop(vm_names=vm_names)
 
 
+@lago.plugins.cli.cli_plugin(help='Export virtual entities')
+@lago.plugins.cli.cli_plugin_add_argument(
+    '--entities-names',
+    '-e',
+    help='Name of the entities to export. If no name is specified, export all'
+    'the entities in this prefix',
+    metavar='ENTITY_NAME',
```
Comment (nvgoldin) [inline lago/cmd.py]: 1. about the meaning of 'entity' - from reading this I'm not sure what it does. 
2. I think here is the place to put more docs about what this command does:
``it exports all VMs into a ..., if standalone is used ....`` etc

File: lago/cmd.py [hunk 2]
```diff
@@ -271,6 +271,39 @@ def do_stop(prefix, vm_names, **kwargs):
...
+    '--standalone',
+    '-s',
+    help='If not specified, export a layered image',
+    action='store_true'
+)
+@lago.plugins.cli.cli_plugin_add_argument(
+    '--compress',
+    '-c',
+    help='If specified, compress the exported images',
+    action='store_true'
```
Comment (nvgoldin) [inline lago/cmd.py]: might be better to use 'store' and give it a default of ``True`` so it could be configurable in lago.conf
(we can't configure 'store_true' ATM)

File: lago/export.py [hunk 3]
```diff
@@ -0,0 +1,239 @@
...
+        return cls(dst, disk_type, disk, do_compress, *args, **kwargs)
+
+    @abstractmethod
+    def export(self):
+        """
+        This method will handle the export process and should
+         implemented in
+        each subclass.
+        """
+        raise NotImplementedError
```
Comment (nvgoldin) [inline lago/export.py]: better add a message here to the exception('abstract method' etc)

File: lago/export.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        if self.standalone:
+            # Consolidate the layers and base image
+            utils.qemu_rebase(target=self.dst, backing_file="")
+        else:
+            if len(self.src_qemu_info) > 2:
+                LOGGER.error(
+                    'Layered export is currently supported for one '
+                    'layer only.  You can try to use Standalone export.'
+                )
+                sys.exit(1)
```
Comment (nvgoldin) [inline lago/export.py]: why exit here? better throw an exception and let it propagate(like in the rest of the code).

File: lago/export.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+        with LogTask('Exporting disk {} to {}'.format(self.name, self.dst)):
+            with utils.RollbackContext() as rollback:
+                rollback.prependDefer(
+                    shutil.rmtree, self.dst, ignore_errors=True
+                )
+                self.copy()
+                self.sparse()
+                self.rebase()
+                self.calc_sha('sha1')
+                self.calc_sha('sha512')
```
Comment (nvgoldin) [inline lago/export.py]: do we always need the two hashes?

File: lago/export.py [hunk 6]
```diff
@@ -0,0 +1,239 @@
...
+                self.update_lago_metadata()
+                self.write_lago_metadata()
+                self.compress()
+                rollback.clear()
+
+
+class FileExportManager(DiskExportManager):
```
Comment (nvgoldin) [inline lago/export.py]: missing docstrings

File: lago/export.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+    def __init__(self, dst, disk_type, disk, do_compress, *args, **kwargs):
+        super(FileExportManager, self).__init__(
+            dst, disk_type, disk, do_compress
+        )
+
+    def export(self):
```
Comment (nvgoldin) [inline lago/export.py]: docstrings

File: lago/plugins/vm.py [hunk 8]
```diff
@@ -174,6 +174,29 @@ def vnc_port(self, *args, **kwargs):
         """
         pass
 
+    def export_disks(
```
Comment (nvgoldin) [inline lago/plugins/vm.py]: docstrings

File: lago/utils.py [hunk 9]
```diff
@@ -561,3 +562,146 @@ def _add_subparser_to_cp(cp, section, actions, incl_unset):
         cp.set(section, var, str(action.default))
     if len(cp.items(section)) == 0:
         cp.remove_section(section)
+
+
+def with_validation(failure_msg):
```
Comment (nvgoldin) [inline lago/utils.py]: I think this decorator is problematic:
1. Use-case: it abstracts away the ability to handle the exceptions inside each function.
2. It relays on 'failure_on_error' to be defined on the decorated function, but it does not enforce setting it, so I would argue its less pythonic: explicit is better than implicit.
3. It relays that the function always returns the return code and specifically that it returns CommandStatus.

I think just updating 'run_command' with those parameters would be simpler: i.e.
```python
result = run_command(cmd=cmd, fail_on_error=True)
```

File: lago/utils.py [hunk 10]
```diff
@@ -561,3 +562,146 @@ def _add_subparser_to_cp(cp, section, actions, incl_unset):
...
+        object: if backing_chain == True then a list of dicts else a dict
+    """
+
+    cmd = ['qemu-img', 'info', '--output=json', path]
+
+    if backing_chain:
+        cmd.insert(-1, '--backing-chain')
+
+    result = run_command(cmd)
+    if result and fail_on_error:
```
Comment (nvgoldin) [inline lago/utils.py]: the decorator is not used on purpose?

File: lago/utils.py [hunk 11]
```diff
@@ -561,3 +562,146 @@ def _add_subparser_to_cp(cp, section, actions, incl_unset):
...
+
+
+@with_validation('Failed to run qemu-img rebase')
+def qemu_rebase(target, backing_file, safe=True):
+    """
+    changes the backing file of 'source' to 'backing_file'
+
+    Args:
+        target(str): Path to the source disk
+        backing_file(str): path to the base disk
```
Comment (nvgoldin) [inline lago/utils.py]: worth documenting what happens when ``backing_file``  is empty(as used in the code)

File: lago/utils.py [hunk 12]
```diff
@@ -561,3 +562,146 @@ def _add_subparser_to_cp(cp, section, actions, incl_unset):
...
+        '--format',
+        input_format,
+        '--in-place',
+        input_file,
+    ]
+    result = run_command(cmd)
+
+    return result
+
+
```
Comment (nvgoldin) [inline lago/utils.py]: I think all of the qemu related command should be moved to a different file - maybe something like image_utils.py or making utils itself a module with 'images' inside.  ``utils.py`` is getting too big and out of context.

File: lago/utils.py [hunk 13]
```diff
@@ -561,3 +562,146 @@ def _add_subparser_to_cp(cp, section, actions, incl_unset):
...
+    Args:
+        file_path (str): Path to the file to generate the hash for
+        checksum (str): hash to apply, one of the supported by hashlib, for
+            example sha1 or sha512
+
+    Returns:
+        str: hash for that file
+    """
+
+    sha = getattr(hashlib, checksum)()
```
Comment (nvgoldin) [inline lago/utils.py]: Might consider passing the hashing function directly instead of a string(safer and flexible)

File: lago/virt.py [hunk 14]
```diff
@@ -166,6 +167,32 @@ def virt_path(self, *args):
     def bootstrap(self):
         utils.invoke_in_parallel(lambda vm: vm.bootstrap(), self._vms.values())
 
+    def export_entities(self, entities_names, standalone, dst_dir, compress):
```
Comment (nvgoldin) [inline lago/virt.py]: Missing docstrings

something not critical - but I think 'entities' as  a name doesn't mean anything, it would be hard to deduce from the name only what this function does.

File: lago/virt.py [hunk 15]
```diff
@@ -166,6 +167,32 @@ def virt_path(self, *args):
...
+                    running_ents.append(ent)
+            except KeyError:
+                LOGGER.error('Entity {} does not exist'.format(name))
+                sys.exit(1)
+
+        if running_ents:
+            for ent in running_ents:
+                LOGGER.error('{} must be off'.format(ent.name()))
+            sys.exit(1)
+        # TODO: run the export task in parallel
```
Comment (nvgoldin) [inline lago/virt.py]: just wondering - what shared-resources prevent it ATM from running in parallel?

File: tests/functional/common.bash [hunk 16]
```diff
@@ -20,6 +20,20 @@ VERBS=(
...
+          --out-format flat \
+          status
+    )
+    echo "$status" | grep -q "status: up"
+    if [[ $? -ne 0 ]]; then
+        return 0
+    else
+        return 1
+    fi
+}
```
Comment (nvgoldin) [inline tests/functional/common.bash]: you need to add the verb in ``common.bash`` also in https://github.com/lago-project/lago/blob/master/tests/functional/common.bash#L3

File: tests/functional/export.bats [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+    local workdir="$FIXTURES"/workdir
+    local suite="$FIXTURES"/suite.json
+
+    rm -rf "$WORKDIR" "$STORE"
+
+    # This is needed to be able to run inside mock, as it uses some temp files
+    # and that is not seamlesly reachable from out of the chroot by
+    # libvirt/kvm
+    export BATS_TMPDIR BATS_TEST_DIRNAME
+    export LIBGUESTFS_DEBUG=1 LIBGUESTFS_TRACE=1
```
Comment (nvgoldin) [inline tests/functional/export.bats]: can be removed - see https://github.com/lago-project/lago/pull/391

File: tests/functional/export.bats [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+    echo "$content" > "dummy_file"
+
+    helpers.run_ok "$LAGOCLI" \
+        copy-to-vm \
+        "$VM_NAME" \
+        dummy_file \
+        /root/dummy_file_inside
+
+    # Just to be sure that the file was written to disk
+    sleep 2
```
Comment (nvgoldin) [inline tests/functional/export.bats]: as we extract files with SCP now, I think this is not needed.

File: tests/functional/export.bats [hunk 19]
```diff
@@ -0,0 +1,167 @@
...
+    helpers.run_nook "$LAGOCLI" \
+        "export" \
+        --standalone \
+        --dst-dir "$STAND_ALONE_EXPORT_DIR"
+}
+
+@test "export.stop" {
+    cd "$FIXTURES"
+
+    "$LAGOCLI" shell "$VM_NAME" "cd /sbin && ./poweroff"
```
Comment (nvgoldin) [inline tests/functional/export.bats]: why is this needed before export? and if so - it should be documented in the CLI.

File: tests/functional/export.bats [hunk 20]
```diff
@@ -0,0 +1,167 @@
...
+    sleep 5
```
Comment (nvgoldin) [inline tests/functional/export.bats]: can it be less than 5?

File: tests/functional/export.bats [hunk 21]
```diff
@@ -0,0 +1,167 @@
...
+    cd "$FIXTURES"
+
+    helpers.run_ok common.is_stopped $WORKDIR
+    helpers.run_ok "$LAGOCLI" \
+        "export" \
+        --standalone \
+        --dst-dir "$STAND_ALONE_EXPORT_DIR"
+}
+
+@test "export.export_layred: layred images export" {
```
Comment (nvgoldin) [inline tests/functional/export.bats]: s/layred/layered

File: tests/functional/export.bats [hunk 22]
```diff
@@ -0,0 +1,167 @@
...
+        "$suite"
+}
+
+@test "export.test_exported_env: testing standalone export" {
+    cd $FIXTURES
+    local dummy_file_content=$(cat dummy_file)
+    cd "$STAND_ALONE_EXPORT_DIR"
+
+    helpers.run_ok "$LAGOCLI" start
+    dummy_file_inside_content=$("$LAGOCLI" shell "$VM_NAME" "cat /root/dummy_file_inside")
```
Comment (nvgoldin) [inline tests/functional/export.bats]: not directly related - but just emphasises the fact we need to fix https://github.com/lago-project/lago/issues/409 as all of these sort of checks might fail too.
btw one way to overcome this is to generate the dummy file with N lines and compare only its 'tail -N'

**Comment by nvgoldin** [comment]
Great job!
2 more things:
1. Might be worth to add to the PR description few running commands examples(and they can later be added to the docs), so anyone who wants(like me:) to test it would know how to run it easily. 

2. Major issue I have is with the tests, currently it can take up to 700 seconds to run the functional tests in CI, I think we cannot afford to increase the running time any more, it makes developement and release process a pain. So first question is how long does it take?
second - I think you should merge the standalone export tests into other existing tests, you already have everything in there(the dummy file, etc) so we could skip the expensive operation of running 'init' again. About the layered images test - need to check how much time does it take, this might also be possible to merge into the existing suite(needs more checking though)

File: lago/prefix.py [hunk 23]
```diff
@@ -493,7 +493,17 @@ def _handle_file_disk(self, disk_spec):
         if url:
             disk_spec['path'] = self._retrieve_disk_url(url, disk_path)
         else:
-            disk_spec['path'] = disk_path
+            # Create a copy of the file or use the original
+            if disk_spec.get('make_a_copy'):
+                LOGGER.info("Making a copy")
+                dest_path = self._generate_disk_path(
+                    os.path.basename(disk_spec['path'])
+                )
```
Comment (nvgoldin) [inline lago/prefix.py]: not sure - would we have a problem here with >1 disks from the same image and ``'make_a_copy=true'``?

File: lago/export.py [hunk 24]
```diff
@@ -0,0 +1,239 @@
...
+        """
+        raise NotImplementedError
+
+    def copy(self):
+        """
+        Copy the disk using cp in order to preserves the 'sparse'
+        structure of the file
+        """
+        with LogTask('Copying disk'):
+            utils.cp(self.src, self.dst)
```
Comment (mykaul) [inline lago/export.py]: Does that copy with sparseness? If not (and even if it does), we should probably run 'virt-sparsify' on the image first. It takes few seconds, and reduces nicely (by up to several hundreds of megs) our disks. virt-sparsify is better than plain 'cp' with sparseness preserveness, as it 'compacts' the image as well - sparsifies marked-as-deleted-but-not-deleted blocks.

File: lago/export.py (hunk 4 above)
Comment (gbenhaim) [inline lago/export.py]: I think that printing a huge stack trace to the screen is a bad user user experience, and should be used only when an actual exception was occurred. IMO when dealing with a known issue it's better to print an explanation on the error and suggest a way to fix it.

File: lago/export.py (hunk 5 above)
Comment (gbenhaim) [inline lago/export.py]: I followed the convention that is being used in Lago Images, even though we don't really use 'sha512'. The verification of the image is done with  'sha128'. If we decide to drop 'sha512' in the export feature we also need to drop it from Lago Images (I don't mind to drop it). 
What do you think ?

File: tests/functional/export.bats (hunk 19 above)
Comment (gbenhaim) [inline tests/functional/export.bats]: Clean shutdown isn't needed before exporting, you can just stop the env using 'lago stop'.
In the case of the functional test, if not shutting down the machine properly the 'dummy_file_insdie' isn't written to disk.

File: lago/prefix.py (hunk 23 above)
Comment (gbenhaim) [inline lago/prefix.py]: Can you give an example ?

Comment (nvgoldin) [inline lago/prefix.py]: something like:
```yaml
domains:
  dom1:
..
    disks:
      - template_name: disk1
        path: '$PWD/disk1.qcow2'
        make_a_copy: True
        type: file
        ....

  dom2:
..
    disks:
      - template_name: disk1
        path: '$PWD/disk1.qcow2'
        make_a_copy: True
        type: file
        ....
```

Or even:
```yaml
domains:
  dom1:
..
    disks:
      - template_name: disk1
        path: '$PWD/disk1.qcow2'
        make_a_copy: True
        type: file
        ....
    - template_name: disk2
        path: '$PWD/disk1.qcow2'
        make_a_copy: True
        type: file
....
```

The idea is to check that the copy we make has a unique name.

File: lago/export.py (hunk 5 above)
Comment (nvgoldin) [inline lago/export.py]: I think if its functional with just one of them, we should keep that way and open an issue to order the to decide what to use. 
If for some reason its not functional with both, even though we just use sha128, then we need to fix that. Maybe just add some dummy hash for now? (i.e. 'not_used_yet). As calculating the hash is relatively expensive.

File: lago/export.py (hunk 4 above)
Comment (nvgoldin) [inline lago/export.py]: I agree about the user-experience, but I don't think this is easily solved by ``sys.exit``:
1. The rest of the code(I think) doesn't do that, which will cause inconsistencies when trying to debug issues. 
2. The 'rebase' command is not a direct CLI command, other functions may call it in different scenarios, not even necessarily as in this PR, so there will be no way to know if they are responsible for a wrong call, or it exits as expected here. 
3. We might have right now(didn't check) or have in the future cleanup code that should be called before the program terminates. Simply exiting the program abruptly might prevent it from running (python raises ``SystemExit`` exception on ``sys.exit`` so I think 'finally' blocks up the stracktrace should be executed, but other stuff might not)
4. Following 3- I'm not sure how python will behave as far as memory cleanup and object releases when issued sys.exit(1), I would guess right now its not a big issue, but might be.
5. There is no logging of the actual exception and its stacktrace. 

If you still want to change this user behaviour, an alternative that comes to mind:

Add in the export function(the direct call the user makes) in  ``cmd.py`` something like:
```python
try:
    ... do your stuff here
catch (your expected exception list) as e:
   LOGGER.info(e.message)
   LOGGER.exception(e)
   sys.exit(1)
```
I can't say I'm 100% pleased with that, as we still swallow the exception(or more correctly raise a different one - ``SystemExit``, instead of the original one), but at least it will be logged to ```lago.log``` and done as close as possible 'up' to the stacktrace, so whoever wanted to clean-up his mess had a chance to do so. 


Further ahead: we can do the same technique for all exception, maybe add a flag that will control whether to print the stacktrace, if set to false, print the message to stdout, and log the entire exception to lago.log. Alternatively, we can control the depth of the printed stacktrace with ```traceback.print_tb```

File: tests/functional/export.bats (hunk 19 above)
Comment (nvgoldin) [inline tests/functional/export.bats]: 👍 
Could also add a check before running export that verifies the VMs are down.

File: lago/utils.py (hunk 13 above)
Comment (gbenhaim) [inline lago/utils.py]: What do you mean ? Can you give an example ?

Comment (nvgoldin) [inline lago/utils.py]: change:
```python
def get_hash(file_path, checksum='sha1'):
```
to:
```
def get_hash(file_path, checksum=hashlib.sha1):
```
So if the caller wants a different function, he would pass it directly, so you don't need to do 'string' -> function conversions.

File: tests/functional/export.bats (hunk 19 above)
Comment (gbenhaim) [inline tests/functional/export.bats]: It exists in virt.py, also added a functional test to check it.

File: lago/virt.py (hunk 15 above)
Comment (gbenhaim) [inline lago/virt.py]: Not something that I can think on for this moment, but let keep things as simple as we can for now, and improve later.

File: tests/functional/export.bats (hunk 18 above)
Comment (gbenhaim) [inline tests/functional/export.bats]: I don't see how it relates, I'm using 'copy-to-vm' which always used ssh, am I missing something?

File: lago/prefix.py (hunk 23 above)
Comment (gbenhaim) [inline lago/prefix.py]: I can't see a situation where one disk is being used by two vms, nor used twice or more times by one vm.

File: lago/export.py (hunk 24 above)
Comment (gbenhaim) [inline lago/export.py]: This copy method is based the os' "cp", which indeed preserve the sparseness  of the file. After the copy I'm running virt-sparsify in "--in-place" mode, which doesn't need large amounts of temporary space.

File: lago/export.py (hunk 5 above)
Comment (gbenhaim) [inline lago/export.py]: for the meanwhile I removed sha512 calculation, as it takes a lot of time and not really being used.

File: tests/functional/export.bats (hunk 18 above)
Comment (nvgoldin) [inline tests/functional/export.bats]: I meant the ``sleep 2``, as we user SCP to extract the files, it should have no affect.

File: tests/functional/export.bats (hunk 19 above)
Comment (nvgoldin) [inline tests/functional/export.bats]: 👍

File: lago/virt.py (hunk 15 above)
Comment (nvgoldin) [inline lago/virt.py]: 👍

File: lago/export.py [hunk 25]
```diff
@@ -203,20 +215,31 @@ def export(self):
...
 
 
 class FileExportManager(DiskExportManager):
+    """
+    FileExportManager is responsible exporting images of type file and empty.
+
+    Attributes:
+        standalone (bool): If true, create a new image which is the result of
+         merging all the layers of src (the image that we want to export).
+         src_qemu_info (dict): Metadata on src which was generated by qemu-img.
```
Comment (nvgoldin) [inline lago/export.py]: I think you need to keep the same indentation, otherwise it would not format nicely with sphinix:
```python
"""
Attributes:
    standalone (bool): long line ...
        more of that long line
    othervar (str): blah
"""
```

File: lago/utils.py [hunk 26]
```diff
@@ -564,37 +564,14 @@ def _add_subparser_to_cp(cp, section, actions, incl_unset):
...
-                raise RuntimeError(
-                    'msg: {}\nfunction: {}\nsubprocess msg: '
-                    '"{}"\nargs: {}\nkwargs: {}'.format(
-                        failure_msg, func.__name__, result.err, args, kwargs
-                    )
-                )
-            return result
-
-        return wrapped
+def run_command_with_validation(
```
Comment (nvgoldin) [inline lago/utils.py]: nice!

File: lago/utils.py [hunk 27]
```diff
@@ -614,62 +591,63 @@ def get_qemu_info(path, backing_chain=False, fail_on_error=True):
...
 
-@with_validation('Failed to run qemu-img rebase')
-def qemu_rebase(target, backing_file, safe=True):
+def qemu_rebase(target, backing_file, safe=True, fail_on_error=True):
     """
     changes the backing file of 'source' to 'backing_file'
+    If backing_file is specified as "" (the empty string),
+    then the image is rebased onto no backing file
+    (i.e. it will exist independently of any backing file).
+    (Taken from qemu-img man page)
```
Comment (nvgoldin) [inline lago/utils.py]: thanks for adding that, makes things clear.

File: lago/virt.py [hunk 28]
```diff
@@ -192,31 +191,33 @@ def virt_path(self, *args):
...
+                )
 
-        if running_ents:
-            for ent in running_ents:
-                LOGGER.error('{} must be off'.format(ent.name()))
-            sys.exit(1)
+        if running_vms:
+            raise utils.LagoUserException(
+                'The following vms must be off:\n{}'.
+                format('\n'.join([_vm.name() for _vm in running_vms]))
```
Comment (nvgoldin) [inline lago/virt.py]: nice

File: lago/virt.py [hunk 29]
```diff
@@ -192,31 +191,33 @@ def virt_path(self, *args):
...
+            )
         # TODO: run the export task in parallel
 
         with LogTask('Exporting disks to: {}'.format(dst_dir)):
-            for ent in entities:
-                ent.export_disks(standalone, dst_dir, compress)
+            for _vm in vms:
```
Comment (nvgoldin) [inline lago/virt.py]: just wondering - why ``_vm``? implies that is a private variable

File: tests/functional/export.bats [hunk 30]
```diff
@@ -132,7 +131,7 @@ export LIBGUESTFS_BACKEND=direct
     cd "$STAND_ALONE_EXPORT_DIR"
 
     helpers.run_ok "$LAGOCLI" start
-    dummy_file_inside_content=$("$LAGOCLI" shell "$VM_NAME" "cat /root/dummy_file_inside")
+    dummy_file_inside_content=$("$LAGOCLI" shell "$VM_NAME" "cat /root/dummy_file_inside" | tail -n1)
```
Comment (nvgoldin) [inline tests/functional/export.bats]: cool - though that file is indeed with more than one line? (otherwise it still might fail)

File: lago/virt.py (hunk 29 above)
Comment (gbenhaim) [inline lago/virt.py]: Otherwise the formatter is complaining about the re-assignment of varaible "vm" . I just needed another name for the variable.

**Comment by gbenhaim** [comment]
ci merge please

**Comment by nvgoldin** [comment]
ci test please

**Comment by gbenhaim** [comment]
ci test please

**Comment by gbenhaim** [comment]
ci merge please
